Panic with v0.1.3

Hi,

I installed v0.1.3 in order to see if it fixes https://github.com/Comcast/trickster/issues/92 but I encountered this.

```
panic: runtime error: index out of range
goroutine 493 [running]:
main.(*PrometheusMatrixEnvelope).cropToRange(0xc0001c06f8, 0x5c06535c, 0x0)
	/go/src/github.com/Comcast/trickster/handlers.go:1014 +0x4ee
main.(*TricksterHandler).originRangeProxyHandler(0xc000062140, 0xc000312640, 0x41, 0xc00014c8a0)
	/go/src/github.com/Comcast/trickster/handlers.go:840 +0x615
created by main.(*TricksterHandler).queueRangeProxyRequest
	/go/src/github.com/Comcast/trickster/handlers.go:660 +0x276
```

Yes `go_gc_duration_seconds{}` made trickster crash

```time=2018-12-05T16:00:22.495898239Z app=trickster caller=trickster/main.go:63 level=info event="application startup" version=0.1.3
time=2018-12-05T16:00:22.496136741Z app=trickster caller=trickster/memory.go:39 level=info event="memorycache setup"
time=2018-12-05T16:00:22.496278442Z app=trickster caller=trickster/metrics.go:47 level=info event="metrics http endpoint starting" address= port=8082
time=2018-12-05T16:00:22.496460743Z app=trickster caller=trickster/main.go:94 level=info event="proxy http endpoint starting" address= port=9090
time=2018-12-05T16:01:00.697780484Z app=trickster caller=trickster/handlers.go:657 level=info event="starting originRangeProxyHandler" cacheKey=7afae6d6003a8f4b6146473157961ddb.8c8344bd560f535d275ce293ffa59420
panic: runtime error: index out of range

goroutine 15 [running]:
main.(*PrometheusMatrixEnvelope).cropToRange(0xc00016ab78, 0x5c06a4bc, 0x0)
        /go/src/github.com/Comcast/trickster/handlers.go:1014 +0x4ee
main.(*TricksterHandler).originRangeProxyHandler(0xc000062100, 0xc000025f90, 0x41, 0xc000066de0)
        /go/src/github.com/Comcast/trickster/handlers.go:840 +0x615
created by main.(*TricksterHandler).queueRangeProxyRequest
        /go/src/github.com/Comcast/trickster/handlers.go:660 +0x276
```

```
time=2018-12-05T16:09:43.448445101Z app=trickster caller=trickster/main.go:63 level=info event="application startup" version=0.1.3
time=2018-12-05T16:09:43.4485415Z app=trickster caller=trickster/memory.go:39 level=info event="memorycache setup"
time=2018-12-05T16:09:43.4488054Z app=trickster caller=trickster/main.go:94 level=info event="proxy http endpoint starting" address= port=9090
time=2018-12-05T16:09:43.4489277Z app=trickster caller=trickster/metrics.go:47 level=info event="metrics http endpoint starting" address= port=8082
time=2018-12-05T16:10:00.666040994Z app=trickster caller=trickster/handlers.go:657 level=info event="starting originRangeProxyHandler" cacheKey=7afae6d6003a8f4b6146473157961ddb.8c8344bd560f535d275ce293ffa59420
time=2018-12-05T16:10:04.061612392Z app=trickster caller=trickster/handlers.go:312 level=debug event=prometheusOriginHttpRequest url="http://prometheus-thanos-query.monitoring.svc.cluster.local:9090/api/v1/query_range?end=1544026200&query=go_gc_duration_seconds&start=1544004600&step=30&timeout=60s" duration=3.395322898s
panic: runtime error: index out of range

goroutine 52 [running]:
main.(*PrometheusMatrixEnvelope).cropToRange(0xc0001fa938, 0x5c06a6d8, 0x0)
        /go/src/github.com/Comcast/trickster/handlers.go:1014 +0x4ee
main.(*TricksterHandler).originRangeProxyHandler(0xc0000ba700, 0xc0002100f0, 0x41, 0xc0001de120)
        /go/src/github.com/Comcast/trickster/handlers.go:840 +0x615
created by main.(*TricksterHandler).queueRangeProxyRequest
        /go/src/github.com/Comcast/trickster/handlers.go:660 +0x276
```

--- a/handlers.go
+++ b/handlers.go
@@ -1031,7 +1031,8 @@ func (pe *PrometheusMatrixEnvelope) cropToRange(start int64, end int64) {
 					break
 				}
 			}
-			if int64(pe.Data.Result[i].Values[len(pe.Data.Result)-1].Timestamp) < start {
+
+			if len(pe.Data.Result[i].Values) == 0 || int64(pe.Data.Result[i].Values[len(pe.Data.Result)-1].Timestamp) < start {
 				seriesToRemove = append(seriesToRemove, i)
 			}
 		}
@@ -1048,8 +1049,9 @@ func (pe *PrometheusMatrixEnvelope) cropToRange(start int64, end int64) {
 					break
 				}
 			}
-			if len(pe.Data.Result[i].Values) > 0 && int64(pe.Data.Result[i].Values[0].Timestamp) > end {
-				if (len(seriesToRemove) > 0 && seriesToRemove[len(seriesToRemove)-1] != i) || len(seriesToRemove) == 0 {
+
+			if len(pe.Data.Result[i].Values) == 0 || int64(pe.Data.Result[i].Values[0].Timestamp) > end {
+				if len(seriesToRemove) == 0 || seriesToRemove[len(seriesToRemove)-1] != i {
 					seriesToRemove = append(seriesToRemove, i)
 				}
 			}

--- a/model_test.go
+++ b/model_test.go
@@ -1,0 +1,194 @@
+package main
+
+import (
+	"reflect"
+	"strconv"
+	"testing"
+
+	"github.com/prometheus/common/model"
+)
+
+func TestPrometheusMatrixEnvelope_CropToRange(t *testing.T) {
+	tests := []struct {
+		before, after PrometheusMatrixEnvelope
+		start, end    int64
+	}{
+		// Case where we trim nothing
+		{
+			before: PrometheusMatrixEnvelope{
+				Data: PrometheusMatrixData{
+					ResultType: "matrix",
+					Result: model.Matrix{
+						&model.SampleStream{
+							Metric: model.Metric{"__name__": "a"},
+							Values: []model.SamplePair{
+								model.SamplePair{1544004600, 1.5},
+							},
+						},
+					},
+				},
+			},
+			after: PrometheusMatrixEnvelope{
+				Data: PrometheusMatrixData{
+					ResultType: "matrix",
+					Result: model.Matrix{
+						&model.SampleStream{
+							Metric: model.Metric{"__name__": "a"},
+							Values: []model.SamplePair{
+								model.SamplePair{1544004600, 1.5},
+							},
+						},
+					},
+				},
+			},
+			start: 0,
+			end:   1644004600,
+		},
+		// Case where we trim everything (all data is too late)
+		{
+			before: PrometheusMatrixEnvelope{
+				Data: PrometheusMatrixData{
+					ResultType: "matrix",
+					Result: model.Matrix{
+						&model.SampleStream{
+							Metric: model.Metric{"__name__": "a"},
+							Values: []model.SamplePair{
+								model.SamplePair{1544004600, 1.5},
+							},
+						},
+					},
+				},
+			},
+			after: PrometheusMatrixEnvelope{
+				Data: PrometheusMatrixData{
+					ResultType: "matrix",
+					Result:     model.Matrix{},
+				},
+			},
+			start: 0,
+			end:   10,
+		},
+		// Case where we trim everything (all data is too early)
+		{
+			before: PrometheusMatrixEnvelope{
+				Data: PrometheusMatrixData{
+					ResultType: "matrix",
+					Result: model.Matrix{
+						&model.SampleStream{
+							Metric: model.Metric{"__name__": "a"},
+							Values: []model.SamplePair{
+								model.SamplePair{100, 1.5},
+							},
+						},
+					},
+				},
+			},
+			after: PrometheusMatrixEnvelope{
+				Data: PrometheusMatrixData{
+					ResultType: "matrix",
+					Result:     model.Matrix{},
+				},
+			},
+			start: 10000,
+			end:   20000,
+		},
+		// Case where we trim some off the beginning
+		{
+			before: PrometheusMatrixEnvelope{
+				Data: PrometheusMatrixData{
+					ResultType: "matrix",
+					Result: model.Matrix{
+						&model.SampleStream{
+							Metric: model.Metric{"__name__": "a"},
+							Values: []model.SamplePair{
+								model.SamplePair{99, 1.5},
+								model.SamplePair{199, 1.5},
+								model.SamplePair{299, 1.5},
+							},
+						},
+					},
+				},
+			},
+			after: PrometheusMatrixEnvelope{
+				Data: PrometheusMatrixData{
+					ResultType: "matrix",
+					Result: model.Matrix{
+						&model.SampleStream{
+							Metric: model.Metric{"__name__": "a"},
+							Values: []model.SamplePair{
+								model.SamplePair{299, 1.5},
+							},
+						},
+					},
+				},
+			},
+			start: 200,
+			end:   300,
+		},
+		// Case where we trim some off the end
+		{
+			before: PrometheusMatrixEnvelope{
+				Data: PrometheusMatrixData{
+					ResultType: "matrix",
+					Result: model.Matrix{
+						&model.SampleStream{
+							Metric: model.Metric{"__name__": "a"},
+							Values: []model.SamplePair{
+								model.SamplePair{99, 1.5},
+								model.SamplePair{199, 1.5},
+								model.SamplePair{299, 1.5},
+							},
+						},
+					},
+				},
+			},
+			after: PrometheusMatrixEnvelope{
+				Data: PrometheusMatrixData{
+					ResultType: "matrix",
+					Result: model.Matrix{
+						&model.SampleStream{
+							Metric: model.Metric{"__name__": "a"},
+							Values: []model.SamplePair{
+								model.SamplePair{199, 1.5},
+							},
+						},
+					},
+				},
+			},
+			start: 100,
+			end:   200,
+		},
+
+		// Case where we aren't given any datapoints
+		{
+			before: PrometheusMatrixEnvelope{
+				Data: PrometheusMatrixData{
+					ResultType: "matrix",
+					Result: model.Matrix{
+						&model.SampleStream{
+							Metric: model.Metric{"__name__": "a"},
+							Values: []model.SamplePair{},
+						},
+					},
+				},
+			},
+			after: PrometheusMatrixEnvelope{
+				Data: PrometheusMatrixData{
+					ResultType: "matrix",
+					Result:     model.Matrix{},
+				},
+			},
+			start: 200,
+			end:   300,
+		},
+	}
+
+	for i, test := range tests {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			test.before.cropToRange(test.start, test.end)
+			if !reflect.DeepEqual(test.before, test.after) {
+				t.Fatalf("mismatch\nexpected=%v\nactual=%v", test.after, test.before)
+			}
+		})
+	}
+}

